[sw/silicon_creator] shutdown_finalize() updates and hardenings

alphan opened the pull request:
This PR:
* Adds a software reset request to `shutdown_hang()` and modifies the inline assembly to include SRAM scrambling key and reset requests in the loop,
* Updates `shutdown_report_error()` to directly write to UART instead of using `log_printf()`, and
* Removes hamming distance histograms from several comments under `sw/device/silicon_creator`.

alphan commented:
[disassembly](https://gist.github.com/8fc523833b452be2292bdb625437a67d)

mcy commented on sw/device/silicon_creator/lib/shutdown.c:
```diff
@@ -255,6 +257,60 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
+  for (size_t i = 0; i < kMsgPrefixLen; ++i) {
+    // `kMsgPrefixLen` is laundered so that it is loaded to a register at every
+    // iteration.
+    if (i >= launder32(kMsgPrefixLen)) {
```
Why is this necessary? Can't you just do this comparison in the loop? Is this intended to be a funny hardening?

alphan replied:
Kind of. The loop already has this check but this is meant to be a redundant check against printing more than we want to, loading `kMsgPrefixLen` at each iteration. PTAL at `bc56` below:

```
launder32():
/workspace/worktrees/opentitan/boot-policy/sw/device/lib/base/hardened.h:242
    bc56:	47b9                	li	a5,14
shutdown_print_prefix():
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:287
    if (i >= launder32(kMsgPrefixLen)) {
    bc58:	00f57a63          	bgeu	a0,a5,bc6c <shutdown_finalize+0x1bc>
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:290
    abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix[i]);
    bc5c:	00d507b3          	add	a5,a0,a3
    bc60:	0007c783          	lbu	a5,0(a5)
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:284
  for (size_t i = 0; i < kMsgPrefixLen; ++i) {
    bc64:	0505                	addi	a0,a0,1
abs_mmio_write32():
/workspace/worktrees/opentitan/boot-policy/sw/device/lib/base/abs_mmio.h:91
  *((volatile uint32_t *)addr) = value;
    bc66:	cf5c                	sw	a5,28(a4)
shutdown_print_prefix():
/workspace/worktrees/opentitan/boot-policy/build-out/../sw/device/silicon_creator/lib/shutdown.c:284
    bc68:	feb517e3          	bne	a0,a1,bc56 <shutdown_finalize+0x1a6>
```

mcy commented on sw/device/silicon_creator/lib/shutdown.c:
```diff
@@ -255,6 +257,60 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
+ *
+ * @param prefix Prefix to print.
+ */
+static ALWAYS_INLINE void shutdown_print_prefix(const char *prefix) {
```
Would it make more sense for this to take a C string and exit if you hit the maximum prefix length? Since you call this and print32 together, maybe it makes sense to merge them and have the `: 0x` be filled in automatically.

If there's a good reason not to do this, it should be documented.

mcy commented on sw/device/silicon_creator/lib/shutdown.c:
```diff
@@ -269,12 +325,40 @@ SHUTDOWN_FUNC(NO_MODIFIERS, shutdown_report_error(rom_error_t reason)) {
+  is_empty = false;
+  do {
+    is_empty =
+        bitfield_bit32_read(abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET),
```
Why not
```c
while(bitfield_bit32_read(...)) {}
```

alphan replied:
It actually looked worse than this with the abs_mmio_read32 but I split it so that the register is read in the loop body.

alphan commented on sw/device/silicon_creator/lib/shutdown.c:
```diff
@@ -255,6 +257,81 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
+  uint32_t status_reg;
+  do {
+    status_reg = abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET);
+  } while (!bitfield_bit32_read(status_reg, UART_STATUS_TXIDLE_BIT));
```
Checked the UART HWIP spec again: TXIDLE is true when either the FIFO is non-empty or there is a transmission. Thus this loop blocks until all characters are transmitted (may never happen if UART is not transmitting for any reason). Enabling transmission unconditionally was an attempt at addressing this but a limit on the number of iterations would be better. Determining this constant requires some work because CPU clock frequency and baud rate depends on the device and they are link-time constants. If we just define the constant here, we end up with multiplication and division operations so I'll trying adding a constant like `kUartNCOValue` and push an update shortly.

alphan commented:
* Changed the prefix to use a `uint32_t`
* Added `kUartTxFifoCpuCycles` to `device.h` to limit the number of iterations while waiting for UART transmission.

[disassembly](https://gist.github.com/3230ed1245b17ad09f0c1928147d4641)

moidx commented on sw/device/silicon_creator/lib/shutdown.c:
```diff
@@ -269,12 +323,38 @@ SHUTDOWN_FUNC(NO_MODIFIERS, shutdown_report_error(rom_error_t reason)) {
-  log_printf("boot_fault: 0x%x\n", (unsigned int)redacted_error);
+  // Print the error message and the raw life cycle state as reported by the
+  // hardware.
+  shutdown_print(MSG_PREFIX('B', 'F', 'V'), redacted_error);
```
Nice!. One more nit: It would be nice to have these log codes in a central location. Maybe a header file?

alphan replied:
Done

mcy commented on sw/device/silicon_creator/lib/shutdown.c:
```diff
@@ -255,6 +259,57 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
+
+  // Print the hex representation of `val`.
+  const char kHexTable[16] = "0123456789abcdef";
+  // `kHexStrLen` is laundered so that it is loaded to a register at every
```
Is unrolling this loop a problem? That's the only reason I can see laundering being something we care about here, since there's no loop completion check.
